server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1616,6 +1615,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }, listener::onFailure);
     }
 
+    /**
+     * Updates the repository generation that running deletes and snapshot finalizations will be based on for this repository if any such
+     * operations are found in the cluster state while setting the safe repository generation.
+     *
+     * @param state  cluster state to update
+     * @param oldGen previous safe repository generation
+     * @param newGen new safe repository generation
+     * @return updated cluster state
+     */
+    private ClusterState updateRepositoryGenerations(ClusterState state, long oldGen, long newGen) {
+        final SnapshotsInProgress snapshotsInProgress = state.custom(SnapshotsInProgress.TYPE);
+        final String repoName = metadata.name();
+        final List<SnapshotsInProgress.Entry> snapshotEntries;
+        if (snapshotsInProgress == null) {
+            snapshotEntries = List.of();
+        } else {
+            snapshotEntries = new ArrayList<>();
+            for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
+                if (entry.repository().equals(repoName) && entry.repositoryStateId() == oldGen) {
+                    snapshotEntries.add(entry.withRepoGen(newGen));
+                } else {
+                    snapshotEntries.add(entry);
+                }
+            }
+        }
+        final SnapshotDeletionsInProgress snapshotDeletionsInProgress = state.custom(SnapshotDeletionsInProgress.TYPE);
+        final List<SnapshotDeletionsInProgress.Entry> deletionEntries;
+        if (snapshotDeletionsInProgress == null) {
+            deletionEntries = List.of();
+        } else {
+            deletionEntries = new ArrayList<>();
+            for (SnapshotDeletionsInProgress.Entry entry : snapshotDeletionsInProgress.getEntries()) {
+                if (entry.repositoryStateId() == oldGen) {
+                    deletionEntries.add(entry.withRepoGen(newGen));
+                } else {
+                    deletionEntries.add(entry);
```

Comment:
Currently, this would all be situations where we concurrently run stuff before the repo has written anything since it was mounted (so some ops pick up a broken generation, others pick up a good one). All of these situations are somewhat safe from corruption thanks to the fact that the finalization step of writing a new index-N will break for them (at least as safe as the current single operation situation as far as I can tell). We can't really assert any hard truths here I'm afraid (S3 could technically always find some index-N with a weird N at any point ... theoretically).

I mean technically we could also just not allow concurrent operations until after the first write to the repository to be in a spot that is easier to reason about. I rejected that idea to some degree, because:
1. The fact that concurrent operations would actually do extra listings for the last index-N in the unsafe startup scenario makes it more likely that we catch a broken operation. 
2. It requires more logic that mixes repo and snapshot service concerns in a confusing way.

Hope that helps, same as with the other comment on this. I'm happy to spell out all the scenarios I can think of for this here if desired (it's quite a few) :)